content/router.xql

```diff
@@ -465,44 +350,91 @@ declare function router:error-description($description as xs:string, $line as xs
  : in the third argument of error().
  :)
-declare function router:send($code as xs:integer, $description as xs:string, $value as item()*, $lookup as function(xs:string, xs:integer) as function(*)?) {
-    if ($description = "" and count($value) = 1 and $value instance of map(*) and map:contains($value, "_config")) then
-        let $route := map:get($value, "_config")
-        let $errorHandler := $route('x-error-handler')
-        return
-            (: if an error handler is defined, call it instead of returning the error directly :)
-            if (exists($errorHandler)) then
-                let $fn :=
-                    try {
-                        $lookup($errorHandler, 1)
-                    } catch * {
-                        error($errors:OPERATION, "Error handler function " || $errorHandler || " could not be resolved")
-                    }
+declare %private function router:error ($code as xs:integer, $error as map(*), $lookup as function(xs:string) as function(*)?) {
+    router:log-error($code, $error),
+    (: unwrap error data :)
+    let $route := $error?_request?config
+    let $error := $error?_error
+    return
+        (: if an error handler is defined, call it instead of returning the error directly :)
+        if (exists($route) and map:contains($route, "x-error-handler"))
+        then (
+            try {
+                let $fn := $lookup($route?x-error-handler)
                 return
-                    if (exists($fn)) then
-                        try {
-                            let $response := $fn(map:get($value, "_response"))
-                            return
-                                router:write-response($response, $code, $route)
-                        } catch * {
-                            (: Catch all errors and add the current route configuration to $err:value,
-                            so we can check it later to format the response :)
-                            error($errors:OPERATION, "Failed to execute error handler " || $errorHandler)
-                        }
-                    else
-                        error($errors:OPERATION, "Error handler function " || $errorHandler || " could not be resolved")
-            else
-                router:write-response(map:get($value, "_response"), $code, $route)
-    else (
-        response:set-status-code($code),
-        response:set-header("Content-Type", "application/json"),
-        util:declare-option("output:method", "json"),
-        if ($description = "") then
-            $value
-        else
-            map {
-                "description": $description,
-                "details": if (exists($value) and map:contains($value, "_response")) then map:get($value, "_response") else $value
+                    router:write-response($code, $fn($error), $route)
+            } catch * {
+                let $_error :=
+                    map {
+                        "code": $err:code,
```

Comment:
@line-o There's quite a penalty for changing the error map structure (adjusting the UI components and assuring backward compatibility)

Reply:
At least "description" should contain the entire error string, including location. Adding more properties is fine though.